pandas/core/strings.py

```diff
@@ -502,6 +502,91 @@ def f(x):
     return result, name
 
+def str_extractall(arr, pat, flags=0):
+    """
+    Find all non-overlapping groups in each subject
+    string in the Series using regular expression pat.
+    When each subject string of the Series has exactly
+    one match, extractall(pat) is the same as extract(pat).
+
+    Parameters
+    ----------
+    pat : string
+        Regular expression pattern
+    flags : int, default 0 (no flags)
+        re module flags, e.g. re.IGNORECASE
+
+    Returns
+    -------
+    A Series with one element for each match if there is only
+    one group in pat. Otherwise, a DataFrame with one row 
+    for each match and one column for each group.
+
+    Examples
+    --------
+    A pattern with one group will return a Series.
+    Indices with no matches will not appear in the result.
+
+    >>> S = Series(["a1a2", "b1", "c1"], ["A", "B", "C"])
+    >>> S.str.extractall("[ab](?P<digit>\d)")
+       match
+    A  0        1
+       1        2
+    B  0        1
+    Name: digit, dtype: object
+    
+    A pattern with multiple groups will return a DataFrame.
+    
+    >>> S.str.extractall("(?P<letter>[ab])(?P<digit>\d)")
+            letter digit
+      match             
+    A 0          a     1
+      1          a     2
+    B 0          b     1
+    
+    Optional groups that do not match are NaN in the result.
+    
+    >>> S.str.extractall("(?P<letter>[ab])?(?P<digit>\d)")
+            letter digit
+      match             
+    A 0          a     1
+      1          a     2
+    B 0          b     1
+    C 0        NaN     1
+
+    """
+    from pandas.core.series import Series
+    from pandas.core.frame import DataFrame
+    from pandas.core.index import MultiIndex
+    regex = re.compile(pat, flags=flags)
+    # the regex must contain capture groups.
+    if regex.groups == 0:
+        raise ValueError("This pattern contains no groups to capture.")
+    names = dict(zip(regex.groupindex.values(), regex.groupindex.keys()))
+    columns = [names.get(1 + i, i) for i in range(regex.groups)]
+    match_list = []
+    index_list = []
+    for subject_key, subject in arr.iteritems():
+        if isinstance(subject, compat.string_types):
+            try:
+                key_list = list(subject_key)
+            except TypeError:
+                key_list = [subject_key]
+            for match_i, match_tuple in enumerate(regex.findall(subject)):
+                na_tuple = [
+                    np.NaN if group=="" else group for group in match_tuple]
+                match_list.append(na_tuple)
+                result_key = tuple(key_list + [match_i])
+                index_list.append(result_key)
+    index = MultiIndex.from_tuples(index_list, names=arr.index.names+["match"])
+    if regex.groups == 1:
+        first_list = [tup[0] for tup in match_list]
+        first_col = columns[0]
+        name = first_col if first_col in regex.groupindex else None
+        result = Series(first_list, index, name=name)
+    else:
```

Comment:
@tdhock I think I agree with you here.ok, let's always return a ``DataFrame``. This will entail changing ``.extract`` and putting in API breaking changes. @jorisvandenbossche ?